From 618a5889ebce169d7bfaf792c198b38efb1900ef Mon Sep 17 00:00:00 2001
From: Jacob Abel <thatcadguy@users.noreply.github.com>
Date: Fri, 18 Apr 2014 09:38:56 -0700
Subject: [PATCH] [optimization] Calculate best initial condition for loops

In order to not impact the code too much, we'll simply calculate the correct values of k here so that execution will simply pass through the existing do loops once and be finished.
---
 src/BOPTools/BOPTools_AlgoTools2D.cxx | 5 +++--
 src/IntTools/IntTools_FClass2d.cxx    | 6 ++++++
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/BOPTools/BOPTools_AlgoTools2D.cxx b/src/BOPTools/BOPTools_AlgoTools2D.cxx
index 41234e2..e43f9a1 100644
--- a/src/BOPTools/BOPTools_AlgoTools2D.cxx
+++ b/src/BOPTools/BOPTools_AlgoTools2D.cxx
@@ -54,6 +54,7 @@
 #include <BOPCol_IndexedMapOfShape.hxx>
 #include <BOPTools.hxx>
 
+#include <Standard_Real.hxx>
 
 static 
   Standard_Boolean CheckEdgeLength (const TopoDS_Edge& E);
@@ -289,14 +290,14 @@ static
       aUP1=aUPeriod+aDelta;
       //
       if (u2 > aUP2) {
-        k=1;
+        k = Ceiling((u2 - aUP1)/aUPeriod);
         do {
           aUx=u2-k*aUPeriod;
           iCnt = k++;
         } while (aUx >= aUP1);
       }
       else if (u2 < -aUP2) {
-        k=1;
+        k = Ceiling((-aUP1 - u2)/aUPeriod);
         do {
           aUx=u2+k*aUPeriod;
           iCnt = (k++) + 1;
diff --git a/src/IntTools/IntTools_FClass2d.cxx b/src/IntTools/IntTools_FClass2d.cxx
index 27fb145..a65be9d 100644
--- a/src/IntTools/IntTools_FClass2d.cxx
+++ b/src/IntTools/IntTools_FClass2d.cxx
@@ -45,6 +45,8 @@
 #include <TColStd_DataMapOfIntegerInteger.hxx>
 #include <TColgp_SequenceOfVec2d.hxx>
 
+#include <Standard_Real.hxx>
+
 //=======================================================================
 //function : IntTools_FClass2d:IntTools:_FClass2d
 //purpose  : 
@@ -553,6 +555,7 @@ IntTools_FClass2d::IntTools_FClass2d()
   if (RecadreOnPeriodic) {
     
     if (IsUPer) {
+      uu += Ceiling((Umin - uu)/uperiod)*uperiod;
       if (uu < Umin)
         while (uu < Umin) {
           uu += uperiod;
@@ -566,6 +569,7 @@ IntTools_FClass2d::IntTools_FClass2d()
     }// if (IsUPer) {
     
     if (IsVPer) {
+      vv += Ceiling((Vmin - vv)/vperiod)*vperiod;
       if (vv < Vmin)
         while (vv < Vmin){
           vv += vperiod;
@@ -703,6 +707,7 @@ IntTools_FClass2d::IntTools_FClass2d()
     {
       if (IsUPer)
 	{
+	  uu += Ceiling((Umin - uu)/uperiod)*uperiod;
 	  if (uu < Umin)
 	    while (uu < Umin)
 	      uu += uperiod;
@@ -715,6 +720,7 @@ IntTools_FClass2d::IntTools_FClass2d()
 	}
       if (IsVPer)
 	{
+	  vv += Ceiling((Vmin - vv)/vperiod)*vperiod;
 	  if (vv < Vmin)
 	    while (vv < Vmin)
 	      vv += vperiod;
-- 
1.9.1

